app/restore_to_file.go

```diff
@@ -59,6 +64,78 @@ func RestoreToFileCmd() cli.Command {
 	}
 }
 
+type Restore struct {
+	sync.RWMutex
+	SnapshotName    string
+	RestoreProgress int
+	RestoreError    error
+	LastUpdatedAt   time.Time
+}
+
+func (r *Restore) UpdateRestoreStatus(snapshot string, rp int, re error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.SnapshotName = snapshot
+	r.RestoreProgress = rp
+	r.RestoreError = re
+	r.LastUpdatedAt = time.Now()
+}
+
+func restore(url string) error {
+	restoreObj := &Restore{}
+
+	backupURL := util.UnescapeURL(url)
+	config := &backupstore.DeltaRestoreConfig{
+		BackupURL: backupURL,
+		DeltaOps:  restoreObj,
+		Filename:  BackupFilePath,
+	}
+
+	err := backupstore.RestoreDeltaBlockBackup(config)
+	if err != nil {
+		return err
+	}
+
+	bar := pb.StartNew(100)
+	periodicChecker := time.NewTicker(Retry * time.Second)
+	progress := 0
+
+	for _ = range periodicChecker.C {
+		restoreObj.RLock()
+		if restoreObj.RestoreProgress == 100 {
+			restoreObj.RUnlock()
+			for ; progress < 100; progress++ {
+				bar.Increment()
+				time.Sleep(time.Millisecond * 2)
+			}
+			periodicChecker.Stop()
+			bar.Finish()
+			return nil
+		}
+		if restoreObj.RestoreError != nil {
+			restoreObj.RUnlock()
+			bar.Finish()
+			periodicChecker.Stop()
+			return restoreObj.RestoreError
+		}
+		now := time.Now()
+		diff := now.Sub(restoreObj.LastUpdatedAt)
+		restoreObj.RUnlock()
+
+		if diff.Seconds() > Retry {
```

Comment:
Separate the periodicChecker with progress based timeout. We can refresh every 5 seconds for the progress bar but we won't want to error out just because it didn't finish restoring one block in 5 seconds.